### PR TITLE
Add GPIO Pt1: The refactorating

### PR DESCRIPTION
**Pull request by twoshark**
Refactor and Clean up in preparation for adding in gpio functionality. 
Some words on modeling:
The Application keeps tracks/manages a Cluster and an Echo server
The Cluster tracks/manages a collection of Devices, and manages comms within the cluster
A Device has control methods, metadata, and extends BaseComponent
BaseComponent manages enable/disable, and has a GPIO 
GPIO (wip): exposes pin state and methods to manipulate it

### CODE REVIEW

File: src/cluster/cluster.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+			resp, err := http.Post(currURL, "application/json", bytes.NewBuffer(body))
+			if err != nil {
+				log.Println("WARNING: Failed to POST to Peer: ", c.SlaveDevices[i].String(), currURL)
+				log.Println(err)
+			} else {
+				defer resp.Body.Close()
+				var result string
+				decoder := json.NewDecoder(resp.Body)
+				decoder.Decode(&result)
```
Comment (dtp263) [inline src/cluster/cluster.go]: Use a message object for generic 'cluster update status'

File: src/cluster/cluster.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	err = c.JoinNetwork(masterHostname)
+	if err != nil {
+		log.Println("Failed to Join Network: PANIC")
+		panic(err)
+	}
+}
+
+// JoinNetwork checks if the master exists and joins the network
+func (c *Cluster) JoinNetwork(URL string) error {
+	parsedURL, err := url.Parse("http://" + URL + "/join_network")
```
Comment (dtp263) [inline src/cluster/cluster.go]: This function is getting kind of long. Can we split it up a little? 

It looks like lines 127 - 136 creates-a-join-network-message.

File: src/component/basecomponent.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+//Disable - force this component to an off or safe state and make it unavaible to command
+func (c *BaseComponent) Disable() {
+	c.Enabled = false
+	c.OnState = false
+}
+
+/*CurrentStateSting just for pretty printing the device info */
+func (c *BaseComponent) CurrentStateSting() string {
```
Comment (dtp263) [inline src/component/basecomponent.go]: Typo. But also maybe lets follow the structure used in solenoid.go and implement String() on each object for pretty printing

File: src/component/solenoid.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package component
+
+// Solenoid - base component + solenoid specific metadata
+type Solenoid struct {
+	BaseComponent
+	Voltage int
+	Type    SolenoidType
+}
+
+func (s *Solenoid) open(duration int) {
```
Comment (dtp263) [inline src/component/solenoid.go]: Is there a reason you are holding off from changing the state variable in this object?

File: src/component/solenoid.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+}
+
+func (s *Solenoid) close(delay int) {
```
Comment (dtp263) [inline src/component/solenoid.go]: Same as above

File: src/main.go [hunk 6]
```diff
@@ -105,4 +58,5 @@ func configureEnvironment() {
 	viper.SetDefault("GOFIRE_PORT", 8001)
 	viper.SetDefault("GOFIRE_MASTER_PORT", 8000)
 	viper.SetDefault("GOFIRE_MASTER_HOST", "127.0.0.1")
+	viper.SetDefault("CLUSTER_NAME", "MasterOfHot")
```
Comment (dtp263) [inline src/main.go]: MasterOfHeat*

File: src/component/solenoid.go (hunk 4 above)
Comment (twoshark) [reply src/component/solenoid.go]: WIP party, OnState is likely going to be replaced by an RPIO pin object

File: src/component/solenoid.go (hunk 5 above)
Comment (twoshark) [reply src/component/solenoid.go]: WIP party, OnState is likely going to be replaced by an RPIO pin object

File: src/main.go (hunk 6 above)
Comment (twoshark) [reply src/main.go]: :fire:

**Comment by dtp263** [comment]
Merging but created issues for remaining comments

**Pull request merged**